src/main/kotlin/cacophony/controlflow/FunctionHandler.kt

```diff
@@ -40,8 +41,82 @@ class FunctionHandlerImpl(
     override fun generateCall(
         arguments: List<CFGNode>,
         result: Register?,
+        respectStackAlignment: Boolean,
     ): CFGFragment {
-        TODO("Not yet implemented")
+        val registerArguments = arguments.zip(REGISTER_ARGUMENT_ORDER)
+        val stackArguments = arguments.drop(registerArguments.size).map { Pair(it, Register.VirtualRegister()) }
+
+        val nodes: MutableList<CFGNode> = mutableListOf()
+
+        if (respectStackAlignment) {
+            // we push two copies of RSP to the stack and either leave them both there,
+            // or remove one of them via RSP assignment
+            val oldRSP = Register.VirtualRegister()
+            nodes.add(CFGNode.Assignment(oldRSP, CFGNode.VariableUse(Register.FixedRegister(X64Register.RSP))))
+            for (i in 0..2) {
+                nodes.add(CFGNode.Push(CFGNode.VariableUse(oldRSP)))
+            }
+
+            // in an ideal world we would do something like "and rsp, ~15" or similar; for now this will do
+            // at the very least split the computation of (RSP + stackArguments.size % 2 * 8) % 16
+            // into two cases depending on the parity of stackArguments.size
+            nodes.add(
+                CFGNode.Assignment(
+                    Register.FixedRegister(X64Register.RSP),
+                    CFGNode.Addition(
+                        CFGNode.VariableUse(Register.FixedRegister(X64Register.RSP)),
+                        CFGNode.Modulo(
+                            CFGNode.Addition(
+                                CFGNode.VariableUse(Register.FixedRegister(X64Register.RSP)),
+                                CFGNode.Constant(stackArguments.size % 2 * 8),
+                            ),
+                            CFGNode.Constant(16),
+                        ),
+                    ),
+                ),
+            )
+        }
+
+        // in what order should we evaluate arguments? gcc uses reversed order
+        for ((argument, register) in registerArguments) {
+            nodes.add(CFGNode.Assignment(Register.FixedRegister(register), argument))
+        }
```

Comment:
Shouldn't we use getVariableAllocation to see where variables go?

Reply:
No, as this is a part of a calling convention.